File: scripts/video_to_sprites.py
"""Extract frames from walking video, remove green screen, build sprite sheet."""
import cv2
import numpy as np
from pathlib import Path

VIDEO = r"c:\Users\12858\Downloads\A_cute_chibi_orange_Pochita_fr (1).mp4"
OUT_DIR = Path("d:/chainsawman/assets/images")
FRAME_COUNT = 8          # Number of frames to extract
FRAME_W = 100            # Target display width (px), height auto from aspect

def main():
    cap = cv2.VideoCapture(VIDEO)
    if not cap.isOpened():
        print("ERROR: Cannot open video")
        return

    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    fps = cap.get(cv2.CAP_PROP_FPS)
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    print(f"Video: {width}x{height}, {total_frames} frames, {fps:.1f} fps")

    # Skip first 10% and last 10% (startup/ending artifacts)
    start_frame = int(total_frames * 0.15)
    end_frame = int(total_frames * 0.85)
    usable = end_frame - start_frame
    print(f"Extracting {FRAME_COUNT} frames from frame {start_frame} to {end_frame}")

    # Extract frames
    raw_frames = []
    for i in range(FRAME_COUNT):
        frame_idx = start_frame + int(usable * i / (FRAME_COUNT - 1))
        cap.set(cv2.CAP_PROP_POS_FRAMES, frame_idx)
        ret, frame = cap.read()
        if ret:
            # Convert BGR to RGB
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            raw_frames.append(frame)
            print(f"  Frame {i+1}/{FRAME_COUNT} @ index {frame_idx}")
    cap.release()

    if len(raw_frames) < FRAME_COUNT:
        print(f"WARNING: Only got {len(raw_frames)} frames")
        FRAME_COUNT = len(raw_frames)

    # Process: remove green screen, find unified bounding box
    processed = []
    all_nonzero = []  # for computing unified bbox

    for i, frame in enumerate(raw_frames):
        # Green screen removal
        r, g, b = frame[:,:,0].astype(float), frame[:,:,1].astype(float), frame[:,:,2].astype(float)

        # Green is dominant: g > r*1.2 and g > b*1.2 and g > 80
        mask = (g > r * 1.15) & (g > b * 1.15) & (g > 80)
        alpha = np.where(mask, 0, 255).astype(np.uint8)

        # Feather edges
        kernel = np.ones((3,3), np.uint8)
        alpha_erode = cv2.erode(alpha, kernel, iterations=1)
        alpha_dilate = cv2.dilate(alpha, kernel, iterations=2)
        alpha = cv2.medianBlur(alpha_dilate, 3)

        # RGBA
        rgba = np.dstack([frame, alpha])
        processed.append(rgba)

        # Track non-transparent pixels for bbox
        ys, xs = np.where(alpha > 30)
        if len(ys) > 0:
            all_nonzero.append((ys.min(), ys.max(), xs.min(), xs.max()))

    # Unified bounding box (with padding)
    pad = 8
    y_min = max(0, min(n[0] for n in all_nonzero) - pad)
    y_max = min(height, max(n[1] for n in all_nonzero) + pad)
    x_min = max(0, min(n[2] for n in all_nonzero) - pad)
    x_max = min(width, max(n[3] for n in all_nonzero) + pad)
    bbox_w = x_max - x_min
    bbox_h = y_max - y_min
    print(f"Unified bbox: {x_min},{y_min} → {x_max},{y_max} ({bbox_w}x{bbox_h})")

    # Crop all frames to unified bbox
    cropped = []
    for rgba in processed:
        crop = rgba[y_min:y_max, x_min:x_max]
        cropped.append(crop)

    # Calculate display height from target width, maintaining aspect ratio
    display_h = int(FRAME_W * bbox_h / bbox_w)
    print(f"Display size: {FRAME_W}x{display_h} per frame")

    # Resize and build sprite sheet
    resized_frames = []
    for crop in cropped:
        resized = cv2.resize(crop, (FRAME_W, display_h), interpolation=cv2.INTER_LANCZOS4)
        resized_frames.append(resized)

    # Horizontal sprite sheet
    sprite_w = FRAME_W * FRAME_COUNT
    sprite = np.zeros((display_h, sprite_w, 4), dtype=np.uint8)
    for i, frame in enumerate(resized_frames):
        sprite[:, i*FRAME_W:(i+1)*FRAME_W] = frame

    # Save
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    out_path = OUT_DIR / "pochita_walk.png"
    cv2.imwrite(str(out_path), cv2.cvtColor(sprite, cv2.COLOR_RGBA2BGRA))
    print(f"\nSaved sprite sheet: {out_path} ({sprite_w}x{display_h}, {FRAME_COUNT} frames)")

    # Also save the first frame for reference
    ref_path = OUT_DIR / "pochita_walk_ref.png"
    cv2.imwrite(str(ref_path), cv2.cvtColor(resized_frames[0], cv2.COLOR_RGBA2BGRA))
    print(f"Saved reference frame: {ref_path}")

    print("Done!")

if __name__ == "__main__":
    main()

File: scripts/video_to_sprites.mjs
/* Extract frames from walking video, remove green screen, build sprite sheet. */
import { execSync, execFileSync } from 'child_process';
import { readFileSync, writeFileSync, mkdirSync, readdirSync, unlinkSync, rmdirSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import sharp from 'sharp';
import ffmpegPath from 'ffmpeg-static';

const __dirname = dirname(fileURLToPath(import.meta.url));
const VIDEO = 'c:/Users/12858/Downloads/A_cute_chibi_orange_Pochita_fr (1).mp4';
const TMP = join(__dirname, '..', 'assets', 'images', '.tmp_frames');
const OUT = join(__dirname, '..', 'assets', 'images');
const FRAME_COUNT = 8;
const FRAME_W = 100;

console.log('Step 1: Extract frames from video...');
mkdirSync(TMP, { recursive: true });

// Get video duration via ffmpeg (info is on stderr, redirect to stdout)
const probeOut = execSync(`"${ffmpegPath}" -i "${VIDEO}" -f null - 2>&1`, { encoding: 'utf8' });
const durMatch = probeOut.match(/Duration: (\d+):(\d+):(\d+\.\d+)/);
if (!durMatch) {
  console.error('Could not parse video duration');
  process.exit(1);
}
const duration = parseInt(durMatch[1]) * 3600 + parseInt(durMatch[2]) * 60 + parseFloat(durMatch[3]);
console.log(`  Duration: ${duration.toFixed(2)}s`);

// Extract evenly spaced frames
const skipStart = duration * 0.15;
const skipEnd = duration * 0.85;
const usable = skipEnd - skipStart;
const interval = usable / (FRAME_COUNT - 1);

for (let i = 0; i < FRAME_COUNT; i++) {
  const t = skipStart + i * interval;
  const outFile = join(TMP, `frame_${String(i).padStart(2, '0')}.png`);
  execFileSync(ffmpegPath, [
    '-ss', t.toFixed(3),
    '-i', VIDEO,
    '-frames:v', '1',
    '-q:v', '1',
    '-y',
    outFile
  ], { stdio: 'ignore' });
  console.log(`  Frame ${i+1}/${FRAME_COUNT} @ ${t.toFixed(2)}s`);
}

// Step 2: Process frames - remove green screen, crop, resize
console.log('Step 2: Process frames (green screen removal)...');
const frames = [];
let globalBBox = null; // { top, bottom, left, right }

for (let i = 0; i < FRAME_COUNT; i++) {
  const fPath = join(TMP, `frame_${String(i).padStart(2, '0')}.png`);
  const { data, info } = await sharp(fPath).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  const { width, height } = info;
  const pixels = new Uint8Array(data);

  // Green screen detection
  for (let j = 0; j < pixels.length; j += 4) {
    const r = pixels[j], g = pixels[j+1], b = pixels[j+2];
    // Green is dominant
    if (g > r * 1.12 && g > b * 1.12 && g > 60) {
      pixels[j+3] = 0;
    }
    // Clean up near-green fringe
    else if (g > r * 1.05 && g > b * 1.05 && g > 40) {
      const greenness = Math.min(1, (g - Math.max(r, b)) / 40);
      pixels[j+3] = Math.round(pixels[j+3] * (1 - greenness * 0.7));
    }
  }

  // Find bounding box of non-transparent pixels
  let top = height, bottom = 0, left = width, right = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * 4;
      if (pixels[idx+3] > 20) {
        if (y < top) top = y;
        if (y > bottom) bottom = y;
        if (x < left) left = x;
        if (x > right) right = x;
      }
    }
  }

  if (globalBBox === null) {
    globalBBox = { top, bottom, left, right };
  } else {
    globalBBox.top = Math.min(globalBBox.top, top);
    globalBBox.bottom = Math.max(globalBBox.bottom, bottom);
    globalBBox.left = Math.min(globalBBox.left, left);
    globalBBox.right = Math.max(globalBBox.right, right);
  }

  frames.push({ pixels, width, height });
}

// Add padding
const PAD = 6;
globalBBox.top = Math.max(0, globalBBox.top - PAD);
globalBBox.bottom = Math.min(frames[0].height - 1, globalBBox.bottom + PAD);
globalBBox.left = Math.max(0, globalBBox.left - PAD);
globalBBox.right = Math.min(frames[0].width - 1, globalBBox.right + PAD);

const cropW = globalBBox.right - globalBBox.left + 1;
const cropH = globalBBox.bottom - globalBBox.top + 1;
const displayH = Math.round(FRAME_W * cropH / cropW);
console.log(`  Crop: ${cropW}x${cropH} → Display: ${FRAME_W}x${displayH}`);

// Step 3: Crop, resize, and build sprite sheet
console.log('Step 3: Build sprite sheet...');
const spriteW = FRAME_W * FRAME_COUNT;
const sprite = await sharp({
  create: { width: spriteW, height: displayH, channels: 4, background: { r: 0, g: 0, b: 0, alpha: 0 } }
});

const composites = [];
for (let i = 0; i < FRAME_COUNT; i++) {
  const { pixels, width } = frames[i];
  // Extract cropped region
  const cropRaw = new Uint8Array(cropW * cropH * 4);
  for (let y = globalBBox.top; y <= globalBBox.bottom; y++) {
    const srcOff = (y * width + globalBBox.left) * 4;
    const dstOff = ((y - globalBBox.top) * cropW) * 4;
    cropRaw.set(pixels.subarray(srcOff, srcOff + cropW * 4), dstOff);
  }

  const buf = await sharp(cropRaw, { raw: { width: cropW, height: cropH, channels: 4 } })
    .resize(FRAME_W, displayH, { kernel: 'lanczos3' })
    .png()
    .toBuffer();

  composites.push({ input: buf, left: i * FRAME_W, top: 0 });
}

const walkPath = join(OUT, 'pochita_walk.png');
await sharp({
  create: { width: spriteW, height: displayH, channels: 4, background: { r: 0, g: 0, b: 0, alpha: 0 } }
}).composite(composites).png().toFile(walkPath);

console.log(`  Saved: pochita_walk.png (${spriteW}x${displayH})`);

// Cleanup
for (let i = 0; i < FRAME_COUNT; i++) {
  unlinkSync(join(TMP, `frame_${String(i).padStart(2, '0')}.png`));
}
rmdirSync(TMP);
console.log('Done!');
